snap/quota/quota.go

```diff
@@ -0,0 +1,327 @@
+		for _, subGrp := range grp.SubGroups {
+			if subGrp == grp.Name {
+				return fmt.Errorf("group has circular sub-group reference to itself")
+			}
+		}
+	}
+
+	// check that if this is a sub-group, then the parent group has enough space
+	// to accommodate this new group (we assume that other existing sub-groups
+	// in the parent group have already been validated)
+	if grp.parentGroup != nil {
+		alreadyUsed := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			if child.Name == grp.Name {
+				continue
+			}
+			alreadyUsed += child.MemoryLimit
+		}
+		// careful arithmetic here in case we somehow overflow the max size of
+		// quantity.Size
+		if grp.parentGroup.MemoryLimit-alreadyUsed < grp.MemoryLimit {
+			remaining := grp.parentGroup.MemoryLimit - alreadyUsed
+			return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+		}
+	}
+
+	return nil
+}
+
+// NewSubGroup creates a new sub group under the current group.
+func (grp *Group) NewSubGroup(name string, memLimit quantity.Size) (*Group, error) {
+	subGrp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+		ParentGroup: grp.Name,
+		parentGroup: grp,
+	}
+
+	// check early that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
+
+	return subGrp, nil
+}
+
+// ResolveCrossReferences takes a set of deserialized groups and sets all
+// cross references amongst them using the unexported fields which are not
+// serialized.
+func ResolveCrossReferences(grps map[string]*Group) error {
+	// iterate over all groups, looking for sub-groups which need to be threaded
+	// together with their respective parent groups from the set
+	for name, grp := range grps {
+		if name != grp.Name {
+			return fmt.Errorf("group has name %q, but is referenced as %q", grp.Name, name)
+		}
+
+		// validate the group, assuming it is unresolved
+		if err := grp.validate(); err != nil {
+			return fmt.Errorf("group %q is invalid: %v", name, err)
+		}
+
+		// first thread the parent link
+		if grp.ParentGroup != "" {
+			parent, ok := grps[grp.ParentGroup]
+			if !ok {
+				return fmt.Errorf("missing group %q referenced as the parent of group %q", grp.ParentGroup, grp.Name)
+			}
+			grp.parentGroup = parent
+
+			// make sure that the parent group references this group
+			found := false
+			for _, parentChildName := range parent.SubGroups {
+				if parentChildName == grp.Name {
+					found = true
```

Comment:
break here?

Reply:
done in https://github.com/snapcore/snapd/pull/10198